File: scripts/extract_strategic_pdf.py
#!/Users/edu/.cache/codex-runtimes/codex-primary-runtime/dependencies/python/bin/python3
import json
import re
import sys
from pathlib import Path

from pypdf import PdfReader


def normalize_spaces(text: str) -> str:
    return re.sub(r"\s+", " ", text or "").strip()


def is_heading(line: str) -> bool:
    cleaned = line.strip()
    if len(cleaned) < 4 or len(cleaned) > 120:
      return False
    if re.match(r"^\d+(\.\d+)*\s+.+", cleaned):
      return True
    if cleaned.isupper() and len(cleaned.split()) <= 14:
      return True
    if cleaned.endswith(":") and len(cleaned.split()) <= 12:
      return True
    if cleaned.startswith(("CAPITULO", "CAPÍTULO", "EJE", "LINEA", "LÍNEA", "TEMA", "DIMENSION", "DIMENSIÓN")):
      return True
    return False


def extract_keywords(title: str, content: str) -> list[str]:
    words = re.findall(r"[A-Za-zÁÉÍÓÚÑáéíóúñ]{4,}", f"{title} {content}")
    seen = []
    for word in words:
        normalized = word.lower()
        if normalized not in seen:
            seen.append(normalized)
        if len(seen) >= 12:
            break
    return seen


def main():
    if len(sys.argv) != 2:
        print(json.dumps({"error": "usage: extract_strategic_pdf.py <pdf_path>"}))
        sys.exit(1)

    pdf_path = Path(sys.argv[1])
    if not pdf_path.exists():
        print(json.dumps({"error": "file_not_found"}))
        sys.exit(1)

    reader = PdfReader(str(pdf_path))
    sections = []
    current = None

    for page_index, page in enumerate(reader.pages, start=1):
        raw_text = page.extract_text() or ""
        lines = [normalize_spaces(line) for line in raw_text.splitlines()]
        lines = [line for line in lines if line]

        for line in lines:
            if is_heading(line):
                if current and current["content"].strip():
                    sections.append(current)
                current = {
                    "id": f"chunk-{len(sections) + 1}",
                    "title": line,
                    "content": "",
                    "pageStart": page_index,
                    "pageEnd": page_index,
                }
                continue

            if current is None:
                current = {
                    "id": "chunk-1",
                    "title": f"Sección inicial página {page_index}",
                    "content": "",
                    "pageStart": page_index,
                    "pageEnd": page_index,
                }

            current["content"] = normalize_spaces(f'{current["content"]} {line}')
            current["pageEnd"] = page_index

    if current and current["content"].strip():
        sections.append(current)

    for index, section in enumerate(sections, start=1):
        section["id"] = f"chunk-{index}"
        section["keywords"] = extract_keywords(section["title"], section["content"])

    print(
        json.dumps(
            {
                "pageCount": len(reader.pages),
                "chunks": sections,
            },
            ensure_ascii=False,
        )
    )


if __name__ == "__main__":
    main()

File: lib/ai/extract-strategic-pdf.ts
import { PDFParse } from 'pdf-parse'
import type { StrategicDocument } from './strategic-documents'

type ExtractedStrategicPdf = {
  pageCount: number
  chunks: StrategicDocument['chunks']
}

export async function extractStrategicPdf(buffer: Buffer): Promise<ExtractedStrategicPdf> {
  const pages = await extractPages(buffer)
  const sections: StrategicDocument['chunks'] = []
  let current: StrategicDocument['chunks'][number] | null = null

  for (const page of pages) {
    const lines = page.text
      .split('\n')
      .map(normalizeSpaces)
      .filter(Boolean)

    for (const line of lines) {
      if (isHeading(line)) {
        if (current?.content.trim()) {
          sections.push(current)
        }

        current = {
          id: `chunk-${sections.length + 1}`,
          title: line,
          content: '',
          pageStart: page.pageNumber,
          pageEnd: page.pageNumber,
          keywords: [],
        }
        continue
      }

      if (!current) {
        current = {
          id: `chunk-${sections.length + 1}`,
          title: `Sección inicial página ${page.pageNumber}`,
          content: '',
          pageStart: page.pageNumber,
          pageEnd: page.pageNumber,
          keywords: [],
        }
      }

      current.content = normalizeSpaces(`${current.content} ${line}`)
      current.pageEnd = page.pageNumber
    }
  }

  if (current?.content.trim()) {
    sections.push(current)
  }

  const chunks = sections.map((section, index) => ({
    ...section,
    id: `chunk-${index + 1}`,
    keywords: extractKeywords(section.title, section.content),
  }))

  return {
    pageCount: pages.length,
    chunks,
  }
}

async function extractPages(buffer: Buffer) {
  const parser = new PDFParse({ data: buffer })

  try {
    const result = await parser.getText({ pageJoiner: '' })
    const extractedPages = Array.isArray(result.pages) ? result.pages : []

    return extractedPages.map((page, index) => ({
      pageNumber:
        typeof page?.num === 'number' && Number.isFinite(page.num)
          ? page.num
          : index + 1,
      text: typeof page?.text === 'string' ? page.text : '',
    }))
  } finally {
    await parser.destroy()
  }
}

function normalizeSpaces(text: string) {
  return text.replace(/\s+/g, ' ').trim()
}

function isHeading(line: string) {
  const cleaned = line.trim()

  if (cleaned.length < 4 || cleaned.length > 120) return false
  if (/^\d+(\.\d+)*\s+.+/.test(cleaned)) return true
  if (cleaned === cleaned.toUpperCase() && cleaned.split(/\s+/).length <= 14) return true
  if (cleaned.endsWith(':') && cleaned.split(/\s+/).length <= 12) return true
  if (
    cleaned.startsWith('CAPITULO') ||
    cleaned.startsWith('CAPÍTULO') ||
    cleaned.startsWith('EJE') ||
    cleaned.startsWith('LINEA') ||
    cleaned.startsWith('LÍNEA') ||
    cleaned.startsWith('TEMA') ||
    cleaned.startsWith('DIMENSION') ||
    cleaned.startsWith('DIMENSIÓN')
  ) {
    return true
  }

  return false
}

function extractKeywords(title: string, content: string) {
  const words = `${title} ${content}`.match(/[A-Za-zÁÉÍÓÚÑáéíóúñ]{4,}/g) ?? []
  const seen: string[] = []

  for (const word of words) {
    const normalized = word.toLowerCase()
    if (!seen.includes(normalized)) {
      seen.push(normalized)
    }

    if (seen.length >= 12) break
  }

  return seen
}
